File: lib/models/vendor_feedback.dart
/// Whether the vendor arrived on time — captured on the private feedback
/// form since nothing else in the data model tracks vendor punctuality.
enum OnTimeAnswer { yes, no, notApplicable }

String onTimeToWire(OnTimeAnswer value) => switch (value) {
      OnTimeAnswer.yes => 'yes',
      OnTimeAnswer.no => 'no',
      OnTimeAnswer.notApplicable => 'not_applicable',
    };

OnTimeAnswer? onTimeFromWire(String? value) => switch (value) {
      'yes' => OnTimeAnswer.yes,
      'no' => OnTimeAnswer.no,
      'not_applicable' => OnTimeAnswer.notApplicable,
      _ => null,
    };

/// A couple's private post-booking feedback for a vendor. Never shown to
/// other couples — only the vendor who owns it and admins can read it
/// (enforced server-side). Only the aggregate CRS/badges derived from this
/// are ever public; see [VendorProfile]'s badge fields.
class VendorFeedback {
  final String id;
  final String coupleId;
  final String vendorId;
  final String inquiryId;
  final String? coupleName;
  final int starRating;
  final String? comment;
  final OnTimeAnswer? onTime;
  final bool isFlagged;
  final String? flagReason;
  final DateTime createdAt;

  const VendorFeedback({
    required this.id,
    required this.coupleId,
    required this.vendorId,
    required this.inquiryId,
    this.coupleName,
    required this.starRating,
    this.comment,
    this.onTime,
    this.isFlagged = false,
    this.flagReason,
    required this.createdAt,
  });

  factory VendorFeedback.fromJson(Map<String, dynamic> json) => VendorFeedback(
        id: json['feedback_id'] as String,
        coupleId: json['couple_id'] as String,
        vendorId: json['vendor_id'] as String,
        inquiryId: json['inquiry_id'] as String,
        coupleName: json['couple_name'] as String?,
        starRating: json['star_rating'] as int,
        comment: json['comment'] as String?,
        onTime: onTimeFromWire(json['on_time'] as String?),
        isFlagged: json['is_flagged'] as bool? ?? false,
        flagReason: json['flag_reason'] as String?,
        createdAt: DateTime.parse(json['created_at'] as String),
      );

  Map<String, dynamic> toJson() => {
        'feedback_id': id,
        'couple_id': coupleId,
        'vendor_id': vendorId,
        'inquiry_id': inquiryId,
        'couple_name': coupleName,
        'star_rating': starRating,
        'comment': comment,
        'on_time': onTime == null ? null : onTimeToWire(onTime!),
        'is_flagged': isFlagged,
        'flag_reason': flagReason,
        'created_at': createdAt.toIso8601String(),
      };
}

File: backend/db/models/vendorFeedback.js
const { DataTypes } = require('sequelize');
const sequelize = require('../sequelize');

// Private by design — a couple's star_rating and comment are never surfaced
// to other couples. Only the vendor who owns it and admins can read a row
// (enforced in routes/vendors.js and routes/admin.js, not just hidden in the
// UI). Only aggregate stats derived from this table (see vendorStats.js) are
// ever public.
const VendorFeedback = sequelize.define('VendorFeedback', {
  feedback_id: {
    type: DataTypes.UUID,
    defaultValue: DataTypes.UUIDV4,
    primaryKey: true,
  },
  couple_user_id: {
    type: DataTypes.UUID,
    allowNull: false,
  },
  vendor_id: {
    type: DataTypes.UUID,
    allowNull: false,
  },
  // The specific booked inquiry that made this couple eligible to submit —
  // kept for traceability even though eligibility today is one-per-vendor.
  inquiry_id: {
    type: DataTypes.UUID,
    allowNull: false,
  },
  star_rating: {
    type: DataTypes.INTEGER,
    allowNull: false,
    validate: { min: 1, max: 5 },
  },
  comment: {
    type: DataTypes.TEXT,
    allowNull: true,
  },
  on_time: {
    type: DataTypes.ENUM('yes', 'no', 'not_applicable'),
    allowNull: true,
  },
  is_flagged: {
    type: DataTypes.BOOLEAN,
    allowNull: false,
    defaultValue: false,
  },
  flag_reason: {
    type: DataTypes.STRING,
    allowNull: true,
  },
}, {
  tableName: 'vendor_feedback',
  createdAt: 'created_at',
  updatedAt: false,
  // One feedback submission per couple-vendor pair, enforced at the DB layer
  // — named explicitly so repeated sync({ alter: true }) runs recognize it
  // instead of piling up duplicate indexes (see user.js for why that matters).
  indexes: [
    { unique: true, fields: ['couple_user_id', 'vendor_id'], name: 'vendor_feedback_couple_vendor_unique' },
  ],
});

module.exports = VendorFeedback;
